### PR TITLE
add vanna localized deployment capabilities

### PR DESCRIPTION
**Pull request by hymvp**
# Checklist:

> [!IMPORTANT]  
> Please review the checklist below before submitting your pull request.

- [ ] Please open an issue before creating a PR or link to an existing issue
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I ran `dev/reformat`(backend) and `cd web && npx lint-staged`(frontend) to appease the lint gods

# Description

A tool has been added to enable vanna to have localized deployment capabilities and can use local vector databases, llm, etc.

- [x] New feature (non-breaking change which adds functionality)

# Testing Instructions

I have verified it in the local environment and the function meets expectations

### CODE REVIEW

**Comment by crazywoola** [comment]
Please fix the lint errors.

**Comment by hymvp** [comment]
@crazywoola fixed, please check, thank you！

**Comment by hymvp** [comment]
hi @crazywoola ,could you please review this code ?thank you !

**Comment by hymvp** [comment]
@crazywoola @takatost Conflict resolved，please help review，Thank you！

**Comment by hymvp** [comment]
Currently, several medical customers have the need to deploy Vanna in locally env, using their own large models and db, and they are already in the POC. hope to merge the code as soon as possible. Thank you.

File: docker/docker-compose.yaml [hunk 1]
```diff
@@ -195,6 +195,7 @@ services:
     volumes:
       # Mount the storage directory to the container, for storing user files.
       - ./volumes/app/storage:/app/api/storage
+      - ./volumes/app/chromadb:/app/chromadb
```
Comment (laipz8200) [inline docker/docker-compose.yaml]: Could you please explain this change?

**Review by laipz8200 — CHANGES_REQUESTED** [review]

File: docker/docker-compose.yaml (hunk 1 above)
Comment (hymvp) [reply docker/docker-compose.yaml]: @laipz8200 Local deployment requires a vector database. Here I hope that the vdb data can be persisted, so it is mapped to the ./volumes/app/chromadb directory.

Comment (laipz8200) [reply docker/docker-compose.yaml]: I believe that the deployment of third-party services shouldn't be included in the dify configuration. If it needs to share a vdb with dify, you can apply this configuration in your local environment.

Comment (hymvp) [reply docker/docker-compose.yaml]: hi @laipz8200 ,already deleted it in yaml, please help check，thank you！

File: api/core/tools/provider/builtin/localvanna/tools/localvanna.py [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+
+from vanna.chromadb import ChromaDB_VectorStore
+from vanna.ollama import Ollama
+
+from core.tools.entities.tool_entities import ToolInvokeMessage
+from core.tools.errors import ToolProviderCredentialValidationError
+from core.tools.tool.builtin_tool import BuiltinTool
+
+
+class VannaTool(BuiltinTool):
```
Comment (laipz8200) [inline api/core/tools/provider/builtin/localvanna/tools/localvanna.py:11]: Could you please consider renaming this class to something like LocalVannaTool?

File: api/core/tools/provider/builtin/localvanna/localvanna.yaml [hunk 3]
```diff
@@ -0,0 +1,25 @@
+identity:
+  author: hymvp
+  name: localvanna
+  label:
+    en_US: LocalVanna
+    zh_Hans: LocalVanna
+  description:
+    en_US: The fastest way to get actionable insights from your database just by asking questions.
```
Comment (laipz8200) [inline api/core/tools/provider/builtin/localvanna/localvanna.yaml:8]: The English description doesn’t align with the Chinese version.

File: api/core/tools/provider/builtin/localvanna/localvanna.yaml [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+    zh_Hans: Vanna本地化部署。
+  icon: icon.png
+credentials_for_provider:
+  api_key:
```
Comment (laipz8200) [inline api/core/tools/provider/builtin/localvanna/localvanna.yaml:12]: Just checking, do we still need an API key from Vanna.AI for this local deployment?

File: api/core/tools/provider/builtin/localvanna/localvanna.yaml (hunk 3 above)
Comment (laipz8200) [reply api/core/tools/provider/builtin/localvanna/localvanna.yaml:8]: I suggest clarifying that your implementation only supports Ollama, as Vanna.ai’s local version also supports other providers.

File: api/core/tools/provider/builtin/localvanna/tools/localvanna.yaml [hunk 5]
```diff
@@ -0,0 +1,243 @@
+identity:
+  name: localvanna
+  author: hymvp
+  label:
+    en_US: LocalVanna
+    zh_Hans: LocalVanna
+description:
+  human:
+    en_US: The fastest way to get actionable insights from your database just by asking questions.
```
Comment (laipz8200) [inline api/core/tools/provider/builtin/localvanna/tools/localvanna.yaml:9]: ditto

File: api/core/tools/provider/builtin/localvanna/localvanna.py [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+            VannaTool().fork_tool_runtime(
+                runtime={
+                    "credentials": credentials,
+                }
+            ).invoke(
+                user_id='',
+                tool_parameters={
+                    "model": "chinook",
+                    "db_type": "SQLite",
+                    "url": "https://vanna.ai/Chinook.sqlite",
```
Comment (laipz8200) [inline api/core/tools/provider/builtin/localvanna/localvanna.py:20]: Why is it necessary to connect to Vanna.ai for validation in this local version?

File: api/core/tools/provider/builtin/localvanna/tools/localvanna.py [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+        #########################################################################################
+        # Due to CVE-2024-5565, we have to disable the chart generation feature
+        # The Vanna library uses a prompt function to present the user with visualized results,
+        # it is possible to alter the prompt using prompt injection and run arbitrary Python code
+        # instead of the intended visualization code.
+        # Specifically - allowing external input to the library’s “ask” method
+        # with "visualize" set to True (default behavior) leads to remote code execution.
+        # Affected versions: <= 0.5.5
+        #########################################################################################
+        generate_chart = True
```
Comment (laipz8200) [inline api/core/tools/provider/builtin/localvanna/tools/localvanna.py:133]: It seems that the behavior here conflicts with your comments. Could you please review it?

**Review by laipz8200 — CHANGES_REQUESTED** [review]
Please check my comments.

By the way, I believe it might be better to implement the local version within the existing Vanna.AI Tool.

**Comment by hymvp** [comment]
hi @laipz8200, I will redesign and integrate into the existing vanna.ai tool。

**Comment by hymvp** [comment]
thanks for your suggestion，I will resubmit a new PR

**Comment by E218PQ** [comment]
Need this function very much, do not know to which PR update?Is there any progress?

**Pull request closed without merging**